### PR TITLE
Add MFA and MFA Unit Tests

### PR DESCRIPTION
**Pull request by sheldonvaughn**
PR to add 3 MFA functions and corresponding unit tests.

### CODE REVIEW

**Comment by pliao28** [comment]
Looks mostly good, added some comments of some possible missing content. One last thing is that we should change the version to a pre-release version. I'm not exactly sure how versioning works in PHP here but we can look into it!

File: lib/MFA.php [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    ) {
+        $enrollPath = "auth/factors/enroll";
+
+        if (!isset($type)) {
+            $msg = "Incomplete arguments: Need to specify a type of factor";
+            throw new Exception\UnexpectedValueException($msg);
+        }
+
+        if ($type != "sms" && $type != "totp") {
+            $msg = "Type Parameter must either be 'sms' or 'totp'";
```
[the remark this answers is not in the document]
Comment (sheldonvaughn) [reply lib/MFA.php:35]: Eng mentioned it was okay to leave this out for now, as it's not mentioned in the docs and may or may not get added, depending on customers requesting email factors.

File: lib/Resource/VerificationChallenge.php [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+    public const RESOURCE_ATTRIBUTES = [
+        "challenge",
+        "valid"
+    ];
+
+    public const RESPONSE_TO_RESOURCE_KEY = [
+        "challenge" => "challenge",
+        "valid" => "valid"
+    ];
+}
```
[the remark this answers is not in the document]
Comment (sheldonvaughn) [reply lib/Resource/VerificationChallenge.php:21]: PHP SDK is different, if a 400 is returned we will execute against a different codeblock so this resource won't be used:

    ```
    if ($responseCode >= 400) {
            if ($responseCode >= 500) {
                throw new Exception\ServerException($response);
            } elseif ($responseCode === 401) {
                throw new Exception\AuthenticationException($response);
            } elseif ($responseCode === 403) {
                throw new Exception\AuthorizationException($response);
            } elseif ($responseCode === 404) {
                throw new Exception\NotFoundException($response);
            }

            throw new Exception\BadRequestException($response);
```

File: tests/WorkOS/MFA.php [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+        $this->mockRequest(
+            Client::METHOD_POST,
+            $path,
+            null,
+            $params,
+            false,
+            $result
+        );
+
+        $enrollFactorTotp = $this->mfa->enrollFactor($type, "test", "ricksanchez", "1234567890");
```
[the remark this answers is not in the document]
Comment (sheldonvaughn) [reply tests/WorkOS/MFA.php]: Parameters are nullable - updated tests to use null.

File: tests/WorkOS/MFA.php [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+        $enrollFactorTotpFixture = $this->enrollFactorTotpResponseFixture();
+
+        $this->assertSame($enrollFactorTotpFixture, $enrollFactorTotp);
+    }
```
[the remark this answers is not in the document]
Comment (sheldonvaughn) [reply tests/WorkOS/MFA.php]: Updated test to use Sms function/fixture

File: lib/MFA.php [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+            $params,
+            true
+        );
+
+        if ($type == "totp") {
+            return Resource\AuthenticationFactorTotp::constructFromResponse($response);
+        } elseif ($type == "sms") {
+            return Resource\AuthenticationFactorSms::constructFromResponse($response);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (sheldonvaughn) [reply lib/MFA.php:68]: Added and tested!

File: lib/Resource/VerificationChallenge.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (pliao28) [reply lib/Resource/VerificationChallenge.php:21]: What would run on a 422 error? We'll se that in something like an already verified factor

[the remark this answers is not in the document]
Comment (robframpton) [reply lib/Resource/VerificationChallenge.php:21]: Once we verify this behavior this branch LGTM!

@sheldonvaughn I have zero knowledge of how PHP is published, but for the other SDKs we've been publishing pre-release versions. Do you know if that's possible with PHP?

[the remark this answers is not in the document]
Comment (pliao28) [reply lib/Resource/VerificationChallenge.php:21]: @robframpton we are planning to push all SDK to stable in our talks with Zeno and DX yesterday about release plans for MFA. So I think we can just bump the version here and push to stable EOD Friday for the Monday announcement.

**Comment by pliao28** [comment]
Do you know what the response would return when it's a 422/already verified factor? Otherwise LGTM and we can bump the version.

**Comment by sheldonvaughn** [comment]
> Do you know what the response would return when it's a 422/already verified factor? Otherwise LGTM and we can bump the version.

@pliao28 per the code block shown here https://github.com/workos/workos-php/blob/fd68eece60b10c8c08ee08d043426032b0e55eb6/lib/Client.php#L66-L77:

For a 422, the 'if' logic would evaluate to line 77:
        throw new Exception\BadRequestException($response);
        
So we would throw an exception with the server response.

**Comment by pliao28** [comment]
> 
Ok that so that would return a BadRequestException and would include the response of say `code: already verified factor` and `message: message of already verified factor here` ?

**Comment by sheldonvaughn** [comment]
> 

Yep - here's the exact messaging when I reproduced the error:

Uncaught WorkOS\Exception\BadRequestException: The authentication challenge 'auth_challenge_01FYW4FKBFKY3XD9164W6M5H2P' has already been verified

This is consistent for other errors with our SSO integration for example, if a code is used that has already been used once.

**Comment by pliao28** [comment]
LGTM! Let's bump the version up and we will keep this open until Monday. When we push all the SDKs MFA to stable.

**Pull request merged**